### ISSUE TITLE
[JENKINS-39231] WinSW: Automatically terminate runaway processes in Windows services

### ISSUE THREAD

**Issue by jenkins-infra-bot** [timja-org/jenkins-gh-issues-poc-11-06-a2#10113]
<p>In Jenkins projects we have many users complaining that the slave/agent is "already connected", because they have a runaway slave/agent process. It happens when WinSW gets terminated without executing the process shutdown logic (force kill) or when WinSW fails to terminate the process.</p>

<p>As a part of WinSW 2.0, it would be great to create a logic, which...</p>

<ul>
	<li>records PID of the created process to the disc</li>
	<li>performs status check of the previously spawned process upon restart</li>
	<li>terminates the runaway process if required</li>
</ul>

<p>It can be done via WinSW 2 "plugin"<br/>
Issue: <a href="https://github.com/kohsuke/winsw/issues/125" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/kohsuke/winsw/issues/125</a></p>

**Attachments:**

- [Slave errors](https://issues.jenkins.io/secure/attachment/34659/Slave errors)

---
<details><summary><i>Originally reported by <a title="Oleg Nenashev" href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=oleg_nenashev">oleg_nenashev</a>, imported from: <a href="https://issues.jenkins.io/browse/JENKINS-39231" target="_blank">WinSW: Automatically terminate runaway processes in Windows services</a></i></summary>
<i><ul>
<li><b>assignee</b>: <a title="Oleg Nenashev" href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=oleg_nenashev">oleg_nenashev</a>
<li><b>reported by</b>: oleg_nenashev
<li><b>status</b>: Resolved
<li><b>priority</b>: Major
<li><b>resolution</b>: Fixed
<li><b>resolved</b>: 2017-03-13T16:01:07+00:00
<li><b>votes</b>: 4
<li><b>watchers</b>: 7
<li><b>imported</b>: 2025-11-06
</ul></i>
</details>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=oleg_nenashev">oleg_nenashev</a>:</i>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=mariem_baccar" class="user-hover" rel="mariem_baccar">mariem_baccar</a> I'm pretty sure you're modifying the wrong issue. This is a <b>Feature Request</b> to the <b>Windows Service Wrapper</b>, which has nothing to do with the problem you describe. </p>

<p>If you see the "slave agent is already connected" issue only after the Docker Slaves installation, please comment in  <a href="https://issues.jenkins.io/browse/JENKINS-28492" title="The server rejected the connection: *** is already connected to this master. Rejecting this connection." class="issue-link" data-issue-key="JENKINS-28492"><del>JENKINS-28492</del></a>  or create another issue to Docker Slaves plugin</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=krogan">krogan</a>:</i>
<p>Jenkins master is on 2.32.1<br/>
Master and slaves running Win2012<br/>
The symptoms sound very familiar to a problem where we've had a jenkins slave up.... then we reboot the windows server (slave).<br/>
When the server returns and the slave is automatically started, it hangs around for about 30secs then terminates connection which kills our job.<br/>
We've also witnessed the hosting windows service winsw 1.17 (which auto upgrades to 1.18) bombs out but leaves the java process running.<br/>
The java process is still keeping the slave active to the master for an indiscriminate amount of time (anywhere between 20secs to 2hrs) before eventually dying of its own accord, with no fresh jobs sent or interaction with the windows service.</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=oleg_nenashev">oleg_nenashev</a>:</i>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=krogan" class="user-hover" rel="krogan">krogan</a> Smells like another issue. <br/>
Anyway, makes sense to retest after the WinSW 2 integration. You can install the new Windows service wrapper manually</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=scm_issue_link">scm_issue_link</a>:</i>
<p>Code changed in jenkins<br/>
User: Oleg Nenashev<br/>
Path:<br/>
 src/main/resources/org/jenkinsci/modules/windows_slave_installer/jenkins-slave.xml<br/>
<a href="http://jenkins-ci.org/commit/windows-slave-installer-module/8dcf02da16e7c95c67c7de95fd078089a8ecf8df" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins-ci.org/commit/windows-slave-installer-module/8dcf02da16e7c95c67c7de95fd078089a8ecf8df</a><br/>
Log:<br/>
  <a href="https://issues.jenkins.io/browse/JENKINS-39231" title="WinSW: Automatically terminate runaway processes in Windows services" class="issue-link" data-issue-key="JENKINS-39231"><del>JENKINS-39231</del></a> - Integrate the Runaway Process Killer extension to terminate runaway processes</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=scm_issue_link">scm_issue_link</a>:</i>
<p>Code changed in jenkins<br/>
User: Oleg Nenashev<br/>
Path:<br/>
 core/pom.xml<br/>
 core/src/main/resources/windows-service/jenkins-slave.xml<br/>
 core/src/main/resources/windows-service/jenkins.xml<br/>
 war/pom.xml<br/>
<a href="http://jenkins-ci.org/commit/jenkins/e698d1de41d4311bf5f8b1d2c40b591109e696e2" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins-ci.org/commit/jenkins/e698d1de41d4311bf5f8b1d2c40b591109e696e2</a><br/>
Log:<br/>
  Update Windows Agent Installer to 1.7 and WinSW to 2.0.2 (#2765)</p>

<ol>
	<li>
	<ol>
		<li>
		<ol>
			<li>WinSW changes</li>
		</ol>
		</li>
	</ol>
	</li>
</ol>


<p>The update includes many fixes and improvements, the full list is provided in the <span class="error">&#91;WinSW changelog&#93;</span>(<a href="https://github.com/kohsuke/winsw/blob/master/CHANGELOG.md" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/kohsuke/winsw/blob/master/CHANGELOG.md</a>). There are several issues referenced in Jenkins bugtracker:</p>

<ul>
	<li><a href="https://issues.jenkins.io/browse/JENKINS-22692" title="Jenkins Windows-Slave throwing exception on shutdown causes connection reset issues" class="issue-link" data-issue-key="JENKINS-22692"><del>JENKINS-22692</del></a>(<a href="https://issues.jenkins-ci.org/browse/JENKINS-22692" class="external-link" target="_blank" rel="nofollow noopener">https://issues.jenkins-ci.org/browse/JENKINS-22692</a>) - Connection reset issues when WinSW gets terminated due to the system shutdown</li>
	<li><a href="https://issues.jenkins.io/browse/JENKINS-23487" title="Add support of shared directories mapping to Windows slave services" class="issue-link" data-issue-key="JENKINS-23487"><del>JENKINS-23487</del></a>(<a href="https://issues.jenkins-ci.org/browse/JENKINS-23487)-" class="external-link" target="_blank" rel="nofollow noopener">https://issues.jenkins-ci.org/browse/JENKINS-23487)-</a> Support of shared directories in WinSW</li>
	<li><a href="https://issues.jenkins.io/browse/JENKINS-39231" title="WinSW: Automatically terminate runaway processes in Windows services" class="issue-link" data-issue-key="JENKINS-39231"><del>JENKINS-39231</del></a>(<a href="https://issues.jenkins-ci.org/browse/JENKINS-39231" class="external-link" target="_blank" rel="nofollow noopener">https://issues.jenkins-ci.org/browse/JENKINS-39231</a>) - Enable Runaway Process Killer by default</li>
	<li><a href="https://issues.jenkins.io/browse/JENKINS-39237" title="Add option to update slave.jar in Windows Service Wrapper on startup (when Jenkins runs on HTTPS)" class="issue-link" data-issue-key="JENKINS-39237"><del>JENKINS-39237</del></a>(<a href="https://issues.jenkins-ci.org/browse/JENKINS-39237" class="external-link" target="_blank" rel="nofollow noopener">https://issues.jenkins-ci.org/browse/JENKINS-39237</a>) - Auto-upgrade of JNLP agent versions on the slaves</li>
</ul>


<ol>
	<li>
	<ol>
		<li>
		<ol>
			<li>Windows Agent Installer changes</li>
		</ol>
		</li>
	</ol>
	</li>
</ol>


<ul>
	<li>Adapt the default configurations to pick fixes above</li>
	<li>Slave => Agent renaming where possible</li>
</ul>


<ol>
	<li>
	<ol>
		<li>
		<ol>
			<li>Jenkins core changes</li>
		</ol>
		</li>
	</ol>
	</li>
</ol>


<ul>
	<li>Modify the configuration template, reference advanced options</li>
	<li>Enable Runaway Process Killer by default</li>
</ul>


<ul>
	<li>Update Windows Agent Installer to 1.7</li>
</ul>


<ul>
	<li>Remove the obsolete jenkins-slave.xml file from the core.</li>
</ul>


<p>Now it is within windows-slave-installer</p>

<ul>
	<li>Use the deployed Snapshot for CI</li>
</ul>


<ul>
	<li>Pick the release version of windows-slave-installer-1.7</li>
</ul>

**Comment by jenkins-infra-bot** [comment]
<i>[Originally duplicated by: <a href="https://github.com/timja-org/jenkins-gh-issues-poc-11-06-a2/issues?q=in%3Atitle%20JENKINS-29825">JENKINS-29825</a>]</i>

**Comment by jenkins-infra-bot** [comment]
<i>[Originally related to: <a href="https://github.com/timja-org/jenkins-gh-issues-poc-11-06-a2/issues?q=in%3Atitle%20JENKINS-24155">JENKINS-24155</a>]</i>

**Comment by jenkins-infra-bot** [comment]
<i>[Originally related to: <a href="https://github.com/timja-org/jenkins-gh-issues-poc-11-06-a2/issues?q=in%3Atitle%20JENKINS-28492">JENKINS-28492</a>]</i>

**Comment by jenkins-infra-bot** [comment]
<i>[Originally related to: <a href="https://github.com/timja-org/jenkins-gh-issues-poc-11-06-a2/issues?q=in%3Atitle%20JENKINS-22024">JENKINS-22024</a>]</i>

**Comment by jenkins-infra-bot** [comment]
<i>[Epic: <a href="https://github.com/timja-org/jenkins-gh-issues-poc-11-06-a2/issues?q=in%3Atitle%20JENKINS-38833%20label%3Aepic">JENKINS-38833</a>]</i>
